### DISCUSSION TITLE
Command line problems with SVN

### DISCUSSION THREAD

**Discussion by Khoji** [WinMerge/winmerge#1503] [Q&A]
When I use WinMerge for conflict resolution via the automatic Tortoise SVN configuration it pops up a perfect three-pane comparison that works exactly as it should after the conflicts are resolved and saved. Tortoise calls this with:

`C:\Program Files (x86)\Utils\WinMerge\WinMergeU.exe /e /ub /fm /wl /wr /dl %tname /dm %bname /dr %yname  %theirs %base %mine /o %merged`

However, when I try to duplicate this in a batch file I can't get it to work properly. The windows aren't in the right order and at least one always comes up with an invalid path. I've tried both of the following and a lot of other variants as well:

```
SET DIFF3="C:\Program Files (x86)\Utils\WinMerge\WinMergeU.exe"
svn.exe diff --diff-cmd %DIFF3% -x "-e -ub -fm -wl -wr -dl %tname -dm %bname -dr %yname %theirs %base %mine -o %merged"
```

Also with the quotes differently positioned, to have the file references outside the -x extensions:

`svn.exe diff --diff-cmd %DIFF3% -x "-e -ub -fm -wl -wr -dl %tname -dm %bname -dr %yname" %theirs %base %mine -o %merged
`
Neither works. It always comes up with the subfolder missing from the first path and the word "mine" at the end of it, although the "mine" should be in the third path. The .svn-base file is then in the second path and the conflicted file with the <<<<<<< .mine tags etc. is in the third path.

I've tried to research everywhere to see how I need to configure this command line but I just can't seem to get it right. WinMerge is registered as a 3-way merge tool and, as I noted, everything is working fine when Tortoise SVN calls it as the merge tool. 

Any ideas? I'm stuck.  ;-/

**Comment by sdottaka** [comment]
I am not familiar with the svn command, but I think %theirs, %base, and %mine are variables that can only be used in TortoseSVN.

Also, if there is a conflict mark in the contents of a conflicting file, you can open it in 3pane mode by simply opening it in WinMerge as follows :

`"C:\Program Files (x86)\Utils\WinMerge\WinMergeU.exe" conflictedfile`

**Reply by Khoji** [reply]
> I am not familiar with the svn command, but I think %theirs, %base, and %mine are variables that can only be used in TortoseSVN.

Thanks, but that is what I'm using. Tortoise is the svn.exe file being referenced, which is its command line executable. Also, when the script gets to this call it's because Tortoise has attempted a commit and found a conflict. That results in the call to WinMerge as the merge tool, so it should be passing the pre-populated Tortoise variables as part of the call. That is what is a little puzzing.

**Comment by sdottaka** [comment]
When you had the conflict, did you run the `svn commit` command from the command prompt?
Or did you select the SVN Commit menu item from the Explorer context menu?

If the former, the svn command is included in TortoiseSVN, but is separate from TortoiseSVN; you cannot use TortoiseSVN variables.
If the latter, how are you configuring the Merge tool in TortoiseSVN?

![image](https://user-images.githubusercontent.com/98126/193431671-a206066f-0bde-4e9f-a5b8-33fe5f0195c0.png)

**Reply by Khoji** [reply]
Thanks again. I'm calling commit via the Tortoise svn.exe from the command line.  I showed how WinMerge is configured in Tortoise in my original post (paragraph 2). That is what I based my own call on, since I assumed that should work with Tortoise. 

To make sure that the problem isn't being caused by the commit call as you suggested, I now fabricated a conflict and branched the script to mergetool directly after the conflict was identified by an update call instead. I checked in the source folder and all the necessary conflict files were created, so the source is being handled correctly. Same problem, however. 

Also, since the call is being made via Tortoise with svn.exe, my understanding is that the Tortoise variables should be available if I'm in the folder where the conflict is present. I think I'm going to have to contact Tortoise support and see what they say.

**Comment by sdottaka** [comment]
Anyway, here's an example of opening a conflict file generated by the `svn update` command with WinMerge.

![svn_conflict](https://user-images.githubusercontent.com/98126/193460932-7e35b1be-a436-4363-b3df-e6f29494e337.gif)

**Reply by Khoji** [reply]
I know that works and it looks as though I'm going to have to use it. The problem is that this script setup is for users working on XML based text projects with version control for collaboration. Many of them will be ovrewhelmed by requests to do things like select conflict files themselves. Doing multiple steps like first resolving conflicts and then marking them as resolved will be beyond them -- I know that from support, unfortunately. Their system will be configured by their administrator, and they will invoke version control with a button in their editor toolbar.  

I was hoping to be able to get Tortoise to automatically pass each conflict file to WinMerge and continue to the next file after the conflict was resolved for the first one.  This works great for Git, but I don't  think it's going to work for SVN. I just found this:

> Edit Conflicts
> This command has no CLI equivalent. It invokes TortoiseMerge or an external 3-way diff/merge tool to look at the files involved in the conflict and sort out which lines to use.

And that's the problem. Tortoise simply doesn't have a command for passing a conflict file to a 3-way merge tool with all the necessary parameters. It can do it internally -- that's what the command line in the config is for -- but not from the command line.

Thanks for all your great suggestions. :)

**Reply by Khoji** [reply]
Found a solution: Since Tortoise always writes the conflict files as soon as a conflict is found during a commit, and one of them is always the conflict file with the additional extension ".mine", it's easy to iterate through them and feed them to the merge tool with the .mine extension removed, which gives you the conflict file WinMerge needs. WinMerge halts operation until you save the conflict, then you can perform an _svn.exe resolved <filename>_, and then the next file comes automatically (the %2 argument it picks up at the start is the Tortoise svn.exe CLI tool):

```
:conflict
SET svntool=%2
SET mergetool="C:\Program Files (x86)\Utils\WinMerge\WinMergeU.exe"
cd /D %3
@echo off
setlocal enableextensions enabledelayedexpansion
call :resolve-conflicts ./ *.mine
goto :eof

:resolve-conflicts
set conflicted=
for /r "%~1" %%P in ("%~2") do (
set conflicted=%%~fP
set conflicted=!conflicted:.mine=!
%mergetool% !conflicted!
%svntool% resolved !conflicted!
) 
```
